Libraries/hipSOLVER/syevdx/main.cpp

```diff
@@ -0,0 +1,208 @@
+
+    // Initialize the hipSOLVER Compatibility API by creating a handle.
+    hipsolverDnHandle_t hipsolver_handle;
+    HIPSOLVER_CHECK(hipsolverDnCreate(&hipsolver_handle));
+
+    // Working space variables.
+    int     lwork{}; /* size of workspace in bytes */
+    double* d_work{}; /* device workspace */
+
+    // Search interval variables.
+    int nev{}; /* number of eigenvalues found */
+    // The search interval can be selected to be (vl, vu].
+    int vl{};
+    int vu{};
+    // It's also possible to obtain only the eigenvalues from the il-th to the iu-th found.
+    int il = 1;
+    int iu = 2;
+
+    // Query and allocate working space.
+    HIPSOLVER_CHECK(hipsolverDnDsyevdx_bufferSize(hipsolver_handle,
+                                                  HIPSOLVER_EIG_MODE_VECTOR,
+                                                  HIPSOLVER_EIG_RANGE_I,
+                                                  HIPSOLVER_FILL_MODE_UPPER,
+                                                  n,
+                                                  d_A,
+                                                  lda,
+                                                  vl,
+                                                  vu,
+                                                  il,
+                                                  iu,
+                                                  &nev,
+                                                  d_W,
+                                                  &lwork));
+    HIP_CHECK(hipMalloc(&d_work, lwork));
+
+    // Compute the eigenvalues (written to d_W) and eigenvectors (written to d_A) of matrix A.
+    HIPSOLVER_CHECK(hipsolverDnDsyevdx(hipsolver_handle,
+                                       HIPSOLVER_EIG_MODE_VECTOR,
+                                       HIPSOLVER_EIG_RANGE_I,
+                                       HIPSOLVER_FILL_MODE_UPPER,
+                                       n,
+                                       d_A,
+                                       lda,
+                                       vl,
+                                       vu,
+                                       il,
+                                       iu,
+                                       &nev,
+                                       d_W,
+                                       d_work,
+                                       lwork,
+                                       d_syevdx_info));
+
+    // Check returned info value.
+    int syevdx_info{};
+    HIP_CHECK(hipMemcpy(&syevdx_info, d_syevdx_info, sizeof(syevdx_info), hipMemcpyDeviceToHost));
+
+    int errors{};
+
+    if(syevdx_info < 0)
+    {
+        std::cout << -syevdx_info << "-th parameter is wrong.\n" << std::endl;
+        errors++;
+        goto cleanup;
+    }
+    else if(syevdx_info > 0)
+    {
+        std::cout << "Computing eigenvalues did not converge.\n" << std::endl;
+        errors++;
+        goto cleanup;
+    }
+    {
```

Comment:
Originally there was an `else`, but then the following suggestion was made:
>Something like
>```c++
>if (/* some condition */) {
>  // clean and return
>}
>
>// continue program
>// cleanup
>```
>or
>```c++
>if (/* some condition */) {
>  goto cleanup_and_return;
>}
>
>// continue program
>
>cleanup_and_return:
>// cleanup
>```
>is generally preferred over:
>```c++
>if (/* inverse of some condition*/) {
>  // continue program
>}
>// cleanup
>```
>as it reduces nesting/indentation.

But after choosing Option 2 (goto approach) removing the braces resulted in a `cannot jump from this goto statement to its label` error, because then the jump bypassed the `eps`, `h_one`. `h_minus_one` and `h_zero` variables initialization. So the braces were left to prevent this error from appearing.

Maybe you have some other opinion about this design?

Reply:
Since the braces have to be kept anyway, I don't think you're saving anything on indentation. Using else instead of goto would also make the code more consistent with the syevj_batched example.

Reply:
Okey, should be changed now. BTW, `sygvd` was also changed so it's consistent with that approach.